File: backend/app/schemas/ai.py
from datetime import date, datetime
from typing import List, Optional

from pydantic import BaseModel, Field


class ScheduleRequest(BaseModel):
    start_date: datetime
    end_date: datetime
    company_id: Optional[int] = None


class SuggestedShift(BaseModel):
    employee_id: int
    employee_name: str
    start_time: datetime
    end_time: datetime
    role: str
    reason: Optional[str] = None


class AIResponse(BaseModel):
    shifts: List[SuggestedShift]
    explanation: Optional[str] = None


class AIScoringConfigBase(BaseModel):
    availability_weight: float = 25.0
    skill_match_weight: float = 25.0
    hours_balance_weight: float = 20.0
    rest_margin_weight: float = 15.0
    weekend_balance_weight: float = 10.0
    night_balance_weight: float = 10.0
    preference_weight: float = 5.0
    min_score_threshold: Optional[float] = None


class AIScoringConfigUpdate(BaseModel):
    availability_weight: Optional[float] = None
    skill_match_weight: Optional[float] = None
    hours_balance_weight: Optional[float] = None
    rest_margin_weight: Optional[float] = None
    weekend_balance_weight: Optional[float] = None
    night_balance_weight: Optional[float] = None
    preference_weight: Optional[float] = None
    min_score_threshold: Optional[float] = None
    tenant_id: Optional[int] = None
    clear_min_score_threshold: bool = False


class AIScoringConfigRead(AIScoringConfigBase):
    id: Optional[int] = None
    tenant_id: int
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class AIAssignPreviewRequest(BaseModel):
    week_start: date
    include_open_shifts: bool = True


class AICandidateScoreBreakdown(BaseModel):
    availability_score: float
    skill_match_score: float
    hours_balance_score: float
    rest_margin_score: float
    weekend_balance_score: float
    night_balance_score: float
    preference_score: float


class AICandidateScore(BaseModel):
    employee_id: int
    employee_name: str
    total_score: float
    score_breakdown: AICandidateScoreBreakdown
    flags: List[str] = Field(default_factory=list)


class AIShiftSuggestion(BaseModel):
    shift_id: int
    recommended_employee_id: Optional[int] = None
    recommended_score: Optional[float] = None
    candidates: List[AICandidateScore] = Field(default_factory=list)
    notes: List[str] = Field(default_factory=list)


class AIUnfilledShift(BaseModel):
    shift_id: int
    reasons: List[str] = Field(default_factory=list)


class AIFairnessSummaryItem(BaseModel):
    employee_id: int
    employee_name: str
    recommended_shift_count: int
    recommended_hours: float


class AIAssignPreviewResponse(BaseModel):
    week_start: date
    scoring_config_used: AIScoringConfigRead
    shift_suggestions: List[AIShiftSuggestion]
    unfilled_shifts: List[AIUnfilledShift]
    fairness_summary: List[AIFairnessSummaryItem]


class AIAssignmentInput(BaseModel):
    shift_id: int
    employee_id: int


class AIAssignApplyRequest(BaseModel):
    week_start: date
    assignments: List[AIAssignmentInput]
    apply_target: str


class AIAssignmentApplied(BaseModel):
    shift_id: int
    employee_id: int


class AIAssignmentRejected(BaseModel):
    shift_id: int
    employee_id: int
    reasons: List[str] = Field(default_factory=list)


class AIAssignApplyResponse(BaseModel):
    applied: List[AIAssignmentApplied]
    rejected: List[AIAssignmentRejected]

File: frontend/lib/api/ai.ts
import api from "@/lib/api"

export interface AIScoringConfig {
    id: number | null
    tenant_id: number
    availability_weight: number
    skill_match_weight: number
    hours_balance_weight: number
    rest_margin_weight: number
    weekend_balance_weight: number
    night_balance_weight: number
    preference_weight: number
    min_score_threshold: number | null
    created_at: string
    updated_at: string
}

export interface AICandidateBreakdown {
    availability_score: number
    skill_match_score: number
    hours_balance_score: number
    rest_margin_score: number
    weekend_balance_score: number
    night_balance_score: number
    preference_score: number
}

export interface AICandidate {
    employee_id: number
    employee_name: string
    total_score: number
    score_breakdown: AICandidateBreakdown
    flags: string[]
}

export interface AIShiftSuggestion {
    shift_id: number
    recommended_employee_id: number | null
    recommended_score: number | null
    candidates: AICandidate[]
    notes: string[]
}

export interface AIUnfilledShift {
    shift_id: number
    reasons: string[]
}

export interface AIFairnessItem {
    employee_id: number
    employee_name: string
    recommended_shift_count: number
    recommended_hours: number
}

export interface AIAssignPreviewResponse {
    week_start: string
    scoring_config_used: AIScoringConfig
    shift_suggestions: AIShiftSuggestion[]
    unfilled_shifts: AIUnfilledShift[]
    fairness_summary: AIFairnessItem[]
}

export interface AIAssignPreviewPayload {
    week_start: string
    include_open_shifts: boolean
}

export interface AIAssignmentInput {
    shift_id: number
    employee_id: number
}

export interface AIAssignApplyPayload {
    week_start: string
    assignments: AIAssignmentInput[]
    apply_target: "DRAFT"
}

export interface AIAssignmentRejected {
    shift_id: number
    employee_id: number
    reasons: string[]
}

export interface AIAssignApplyResponse {
    applied: AIAssignmentInput[]
    rejected: AIAssignmentRejected[]
}

export interface AIScoringConfigUpdatePayload {
    availability_weight?: number
    skill_match_weight?: number
    hours_balance_weight?: number
    rest_margin_weight?: number
    weekend_balance_weight?: number
    night_balance_weight?: number
    preference_weight?: number
    min_score_threshold?: number | null
    clear_min_score_threshold?: boolean
}

export async function previewAIAssignments(
    payload: AIAssignPreviewPayload
): Promise<AIAssignPreviewResponse> {
    const res = await api.post("/ai/assign/preview", payload)
    return res.data
}

export async function applyAIAssignments(
    payload: AIAssignApplyPayload
): Promise<AIAssignApplyResponse> {
    const res = await api.post("/ai/assign/apply", payload)
    return res.data
}

export async function getAIScoringConfig(): Promise<AIScoringConfig> {
    const res = await api.get("/settings/ai-scoring")
    return res.data
}

export async function updateAIScoringConfig(
    payload: AIScoringConfigUpdatePayload
): Promise<AIScoringConfig> {
    const res = await api.put("/settings/ai-scoring", payload)
    return res.data
}
